uktrade/great-cms
# GP2 736: Lesson placeholders

This changeset adds the facility to mix 'placeholder' blocks with specified pages in a lesson module/CuratedListPage's list of topics.

Before the `CuratedListPage.topics` `StreamField` contained only a `ListBlock` of `DetailPage`s (via `PageChooserBlock`). Now the `StreamField` contains a `StreamBlock` which can hold either a `DetailPage` (via `PageChooserBlock`) or a new `LessonPlaceholderBlock`

It's a large-ish PR because it:
* involves two schema migrations and a data migration 
* required adding a helper util for to set up test state for a StreamBlock (which is not easily supported via fixtures)
* required amending of a bunch of tests and fixtures
* has a chunk of FE changes too, of course.

 - [X] Ticket exists in Jira  https://uktrade.atlassian.net/browse/GP2-736 

 - [ ] (if is a UI change) Includes screenshot(s) - ideally before and after, but at least after
 - [X] How to test locally, including how to set up appropriate data
 
 
## To note

If you disagree with any of the following, that's cool - let's talk! 

* The data migration has no reverse function - it's a one-way trip. This is a bit riskier if something unexpected is spotted after we hit staging, especially as content is being added, but we do have hourly backups.
* The data migration deletes old page revisions, because the data model no longer matches the older revision's data. I could have updated that during the data migration, but it did not seem to add much value as we are pre-beta. 

![Screenshot 2020-10-19 at 16 49 31](https://user-images.githubusercontent.com/101457/96474850-28109080-122b-11eb-9865-d0eb3351459c.png)
![Screenshot 2020-10-19 at 16 49 43](https://user-images.githubusercontent.com/101457/96474853-2941bd80-122b-11eb-947f-8ed47782befd.png)
![Screenshot 2020-10-19 at 16 49 50](https://user-images.githubusercontent.com/101457/96474858-29da5400-122b-11eb-9a4c-62d22b5dd674.png)







## To Test

* Add a new Lesson/DetailPage and publish it
* On a CuratedListPage / module page, edit its topics to incude the new lesson
* Also try adding a few placeholder blocks, in between lessons, at the start and end of topics, etc
* Publish the module page

* Confirm the module page shows the right number of topics and lessons
* Confirm the lesson pages proceed through 'next lesson' as you'd expect (ie, always taking the user to the next real lesson, skipping placeholders
* If you want to rummage in the DB, confirm that the lesson's topic_block_id matches the topic id from the streamfield (this logic has not been refactored and the topic ID in the CuratedListPage's streamfield is unchanged by the refactor or the data migration

<img width="1571" alt="Screenshot 2020-10-20 at 08 50 59" src="https://user-images.githubusercontent.com/66948936/96556872-723c5500-12b1-11eb-99ac-934426b5aa64.png">
<img width="1394" alt="Screenshot 2020-10-20 at 08 51 05" src="https://user-images.githubusercontent.com/66948936/96556883-76687280-12b1-11eb-8df0-1a383ced8b9a.png">
